database/src/error.rs

```diff
@@ -2,61 +2,148 @@
 
 //---------------------------------------------------------------------------------------------------- InitError
-/// Database errors that occur during initialization.
-///
-/// `BackendError` is an error specifically from the
-/// database backend being used. TODO: this may not
-/// be needed if we can convert all error types into
-/// "generic" database errors.
+/// Errors that occur during ([`Env::open`]).
 #[derive(thiserror::Error, Debug)]
-pub enum InitError<BackendError: Debug> {
-    /// TODO
-    #[error("database PATH is inaccessible: {0}")]
-    Path(std::io::Error),
+pub enum InitError {
+    /// I/O error.
+    #[error("database I/O error: {0}")]
+    Io(#[from] std::io::Error),
+
+    /// The given `Path/File` existed and was accessible,
+    /// but was not a valid database file.
+    #[error("database file exists but is not valid")]
+    Invalid,
+
+    /// The given `Path/File` existed, was a valid
+    /// database, but the version is incorrect.
+    #[error("database file is valid, but version is incorrect")]
+    InvalidVersion,
 
-    /// TODO
-    #[error("{DATABASE_BACKEND} error: {0}")]
-    Backend(BackendError),
+    /// The given `Path/File` existed,
+    /// was a valid database, but it is corrupt.
+    #[error("database file is corrupt")]
+    Corrupt,
 
-    /// TODO
+    /// The database is currently in the process
+    /// of shutting down and cannot respond.
+    ///
+    /// TODO: This might happen if we try to open
+    /// while we are shutting down, `unreachable!()`?
+    #[error("database is shutting down")]
+    ShuttingDown,
+
+    /// A fatal error occurred.
+    ///
+    /// This is for errors that _should_ be unreachable
+    /// but we'd still like to panic gracefully.
     ///
+    /// # Invariant
+    /// If this error is received, all(?) of Cuprate should shutdown.
+    #[error("fatal error: {0}")]
+    Fatal(BoxError),
+
     /// An unknown error occurred.
+    ///
+    /// This is a catch-all for all non-fatal errors.
     #[error("unknown error: {0}")]
-    Unknown(Cow<'static, str>),
+    Unknown(BoxError),
 }
 
 //---------------------------------------------------------------------------------------------------- RuntimeError
-/// Database errors that occur _after_ successful initialization.
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-#[cfg_attr(
-    feature = "borsh",
-    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
-)]
-#[derive(thiserror::Error, Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+/// Errors that occur _after_ successful ([`Env::open`]).
+///
+/// There are no errors for:
+/// 1. Missing tables
+/// 2. (De)serialization
+///
+/// as `cuprate_database` upholds the invariant that:
+///
+/// 1. All tables exist
+/// 2. (De)serialization never fails
+#[derive(thiserror::Error, Debug)]
 pub enum RuntimeError {
-    // TODO: replace string with actual error type.
+    /// The given key already existed in the database.
```

Comment:
There are some errors here that I don't think should be passed back out of the DB, they could still be defined here but they should just panic.

Reply:
Which ones specifically? Should they be removed and panic in the `From` impl?

Reply:
`MapFull`, `PageFull`, `TxMustAbort`, `ReadersFull`, `Corrupt`, `Fatal`, `Unknown`.

IMO the only errors that should be given out from the DB are errors on specific requests i.e. `KeyNotFound` repeating this request with a valid key should then return ok.

Whereas with an error like `Corrupt` the request might have been a valid one but the DB is in an invalid state, if we were to return this error it could cause the rest of Cuprate to enter an invalid state, unless they too were to check DB errors which would cause parts of the DB to leak into places I don't want it to.

How exactly these are handled I don't mind, it would probably be better to remove them and just panic in the from impl (unless these can be fixed like `MapFull`).

Is `InvalidVersion` an invalid Cuprate DB version? If so I think this should be removed from the `RuntimeError` as well and just be an `InitError`